spirv_glsl.cpp

```diff
@@ -6832,6 +6830,9 @@ void CompilerGLSL::emit_subgroup_op(const Instruction &i)
 	case OpGroupNonUniformAny:
 	case OpGroupNonUniformAllEqual:
 		require_extension_internal("GL_KHR_shader_subgroup_vote");
+		// to be able to emulate subgroupAllEqual(T) where T is not bool
+		if (!options.vulkan_semantics)
+			require_extension_internal("GL_KHR_shader_subgroup_ballot");
```

Comment:
you need to track more granular.

Reply:
why?

Reply:
because you often get a situation where one extension defines multiple functionalities that need different subsets of desktop OpenGL extensions to express, the subset of extensions could be bigger if you actually analyze what functionality is actually being used.

For example:

# `GL_KHR_shader_subgroup_basic` (which is implicitly enabled by all the other ones)

If `gl_NumSubgroups` and `gl_SubgroupID` are used at all, then only the Nvidia extension can be used.

Do you want to implicitly enable all of the functionality that `GL_KHR_shader_subgroup_basic` provides and have SPIRV-Cross fail to produce dekstop OpenGL GLSL for other vendors, even when these two special builtins are not being used?

Do you wish to fail to compile a shader that does not use the above and does not use `gl_SubgroupInvocationID` when only `AMD_gcn_shader` is available?

Reply:
Basically I see two sane methods to approach this, either:

A) split the subgroup extensions into more "meta-extensions" and require these when finding an opcode/builtin

B) Keep a list of states for `GL_NV_shader_thread_group, GL_NV_shader_thread_shuffle, GL_ARB_shader_group_vote, GL_NV_gpu_shader5, GL_AMD_gcn_shader, GL_ARB_shader_ballot` and flag them as:
- "not needed" (no opcode or builtin has been encountered which could be expressed in a way requiring this extension)
- "potentially needed" (logical inverse of the above)
- "required" (only way to express a functionality, i.e. `gl_NumSubgroups` forces NV extension)
- "redundant" (another extension which is "required" and the extensions it requires covers this extension 100% in features which will be used)